### PR TITLE
Moving tests to use transactions

### PR DESCRIPTION
**Pull request by dhensby**
Attempting to use transactions for tests to try to increase speed / reduce the need to rebuild all fixtures for each test

---

Depends on #7848

### CODE REVIEW

**Comment by tractorcow** [comment]
Whoops a lot of extra errors here. :)

File: src/Dev/SapphireTest.php [hunk 1]
```diff
@@ -389,17 +391,19 @@ public static function tearDownAfterClass()
...
         static::resetDBSchema();
+
+        static::$fixtureFactory = null;
     }
 
     /**
      * @return FixtureFactory
      */
-    public function getFixtureFactory()
+    public static function getFixtureFactory()
```
Comment (tractorcow) [inline src/Dev/SapphireTest.php]: Semver major change.

File: src/Dev/SapphireTest.php [hunk 2]
```diff
@@ -389,17 +391,19 @@ public static function tearDownAfterClass()
...
     {
-        if (!$this->fixtureFactory) {
-            $this->fixtureFactory = Injector::inst()->create(FixtureFactory::class);
+        if (!static::$fixtureFactory) {
+            static::setFixtureFactory(Injector::inst()->create(FixtureFactory::class));
         }
-        return $this->fixtureFactory;
+        return static::$fixtureFactory;
```
Comment (tractorcow) [inline src/Dev/SapphireTest.php]: Why does this need to be static now? the fixture factory belongs to the class instance. It runs the risk of cross-test pollution.

File: src/Dev/SapphireTest.php [hunk 3]
```diff
@@ -546,6 +549,8 @@ protected function tearDown()
 
         // Call state helpers
         static::$state->tearDown($this);
+
+        static::$tempDB->rollbackTransaction();
```
Comment (tractorcow) [inline src/Dev/SapphireTest.php]: The rollbackTransaction() needs to be in an if() statement that matches the if() on setUp().

`if ($this->shouldSetupDatabaseForCurrentTest($fixtureFiles)) {`

Otherwise you're rolling back more than you're starting transaction.

File: src/ORM/Connect/TempDatabase.php [hunk 4]
```diff
@@ -23,6 +23,8 @@ class TempDatabase
      */
     protected $name = null;
 
+    private $hasStarted = false;
```
Comment (tractorcow) [inline src/ORM/Connect/TempDatabase.php]: protected, and phpdoc. It's not clear that it's a transaction.

File: src/Dev/SapphireTest.php (hunk 2 above)
Comment (dhensby) [reply src/Dev/SapphireTest.php]: So something interesting that I found out when doing this work is that PHPUnit internally creates a new  object instance for every test method. That means the fixtureFactory instance is not shared from test-to-test on the same class. It has to be a static to allow it to be shared across tests.

The alternative is to have a way of loading a fixture without writing to the DB.

**Comment by dhensby** [comment]
OK - I've got the tests passing locally. With this patch the framework test suite passes in 4.5 mins, without it takes 10.2mins - so we are seeing a 56% improvement in speed. It also saved 8MB of memory (on PHP 7.1 - savings on 5.x will be much higher).

Whilst I appreciate the changing of methods to `static` is a semver violation, we should try to explore how we can get this into 4

**Comment by tractorcow** [comment]
Awesome... how much of it is due to the static fixture factory do you think?

**Comment by tractorcow** [comment]
> Whilst I appreciate the changing of methods to static is a semver violation, we should try to explore how we can get this into 4

Just make the new static methods a different name, and deprecate the old ones. ;)

**Comment by tractorcow** [comment]
How do you feel about moving the entire test fixture system into a TestState? I had originally wanted to do this, but it was going to be a huge amount of work. It would reduce the SapphireTest api surface area a great deal. :)

**Comment by dhensby** [comment]
>how much of it is due to the static fixture factory do you think?

I'm not sure - I think it's mostly savings because we aren't trashing, and re-inserting, the fixtures to the DB. I'm sure it's more a DB bottleneck than a YAML parsing one (though it may well be both).

> Just make the new static methods a different name, and deprecate the old ones. ;)

that doesn't stop it being a breaking change. Any classes that have overridden the current method will no longer have the effect they used to, therefore breaking people's code.

> How do you feel about moving the entire test fixture system into a TestState? I had originally wanted to do this, but it was going to be a huge amount of work. It would reduce the SapphireTest api surface area a great deal. :)

This was what I was thinking too... but it does seem like a lot of work too :P - it's an interesting idea.

This PR at the moment was a semi-proof-of-concept to see just what kind of speed up we could get.

**Comment by dhensby** [comment]
OK - moved to a `TestState` and made minimal API changes to `SapphireTest` - I've had to add a few APIs to both `SapphireTestState` and `SapphireTest` to allow the `FixtureTestState` to access the information it needs about tests.

File: src/ORM/Connect/TempDatabase.php [hunk 5]
```diff
@@ -68,6 +73,46 @@ public function isUsed()
...
+            static::getConn()->transactionStart();
+        }
+    }
+
+    /**
+     * Rollback a transaction (or trash all data if the DB doesn't support databases
+     */
+    public function rollbackTransaction()
+    {
+        if (static::getConn()->supportsTransactions()) {
```
Comment (tractorcow) [inline src/ORM/Connect/TempDatabase.php]: Why does this not set hasStarted to false if it supports transactions?

Comment (dhensby) [reply src/ORM/Connect/TempDatabase.php]: So `$hasStarted` is really "do we have the fixtures in the DB" and the only time we don't have the fixtures in the DB is when we clear the DB... it's more "have the tests started" rather than "are we in a transaction"

Comment (dhensby) [reply src/ORM/Connect/TempDatabase.php]: clearAllData actually resets hasStarted so we don't really need it in this method at all

File: src/Dev/State/FixtureTestState.php [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+     * @param SapphireTest $test
+     *
+     * @return bool
+     */
+    protected function testNeedsDB(SapphireTest $test)
+    {
+        $annotations = $test->getAnnotations();
+
+        // annotation explicitly disables the DB
+        if (array_key_exists('useDatabase', $annotations['method'])
```
Comment (tractorcow) [inline src/Dev/State/FixtureTestState.php]: I never even knew this was a thing.

Comment (dhensby) [reply src/Dev/State/FixtureTestState.php]: I saw the PR adding it a while ago - we really should add the annotations to appropriate tests in core tests :)

File: src/Dev/SapphireTest.php [hunk 7]
```diff
@@ -253,31 +281,10 @@ protected function setUp()
...
-            }
-
-            foreach ($fixtureFiles as $fixtureFilePath) {
-                $fixture = YamlFixture::create($fixtureFilePath);
-                $fixture->writeInto($this->getFixtureFactory());
-            }
-
+            /** @var FixtureTestState $fixtureState */
+            $fixtureState = static::$state->getStateByName('fixtures');
+            $this->setFixtureFactory($fixtureState->getFixtureFactory(static::class));
```
Comment (tractorcow) [inline src/Dev/SapphireTest.php]: Is this needed still? Doesn't `this->setFixtureFactory()` just send the factory back to the teststate?

**Comment by tractorcow** [comment]
I think I need to do more testing.

File: src/Dev/SapphireTest.php (hunk 7 above)
Comment (dhensby) [reply src/Dev/SapphireTest.php]: we need to set `$this->fixutureFactory` (property on `SapphireTest`) because some tests access it directly rather than through the helper methods or `getFixtureFactory` :(

File: src/ORM/Connect/TempDatabase.php [hunk 8]
```diff
@@ -183,6 +229,11 @@ public function deleteAll()
      */
     public function resetDBSchema(array $extraDataObjects = [])
     {
+        // pgsql doesn't allow schema updates inside transactions
```
Comment (dhensby) [inline src/ORM/Connect/TempDatabase.php]: TBH, I think this is most DBMS as in MySQL alter tables cause an implicit commit for transactions too...

**Comment by dhensby** [comment]
@chillu can we get some input on this - it's a real shame that this has missed a large number of builds (especially CWP ones) that could have highlighted any edge cases with this fix (as well as saving a lot of build time)

File: src/Dev/State/FixtureTestState.php [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+     * @param SapphireTest $test
+     *
+     * @return string
+     */
+    protected function resolveFixturePath($fixtureFilePath, SapphireTest $test)
+    {
+        // Support fixture paths relative to the test class, rather than relative to webroot
+        // String checking is faster than file_exists() calls.
+        $isRelativeToFile
+            = (strpos('/', $fixtureFilePath) === false)
```
Comment (mikenz) [inline src/Dev/State/FixtureTestState.php]: `int strpos ( string $haystack , mixed $needle [, int $offset = 0 ] )`
Around the wrong way?

Comment (tractorcow) [reply src/Dev/State/FixtureTestState.php]: Ah yes.

I also like strstr() for doing simple is_in_string :P

Comment (dhensby) [reply src/Dev/State/FixtureTestState.php]: Hah, this has been lifted directly from `SapphireTest` - so fixing this may have unintended consequences...

I've pushed up a fix, though

File: src/Dev/State/FixtureTestState.php [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+
+    /**
+     * Called on tear down
+     *
+     * @param SapphireTest $test
+     */
+    public function tearDown(SapphireTest $test)
+    {
+        if ($this->testNeedsDB($test)) {
+            $test::tempDB()->rollbackTransaction();
```
Comment (tractorcow) [inline src/Dev/State/FixtureTestState.php]: Ok, follow up to https://github.com/silverstripe/silverstripe-framework/issues/8182, we need to check if this rollback failed.

If it failed, we need to destroy the database instead. That will cause the next setUp() to run `$tmpDB->build();`. 

Easy fix @dhensby ?

Comment (tractorcow) [reply src/Dev/State/FixtureTestState.php]: You'll need to rebuild fixture factories as well, since those normally only get rebuilt on setuponce.
